File: src/Atlasd/Battlenet/ProtocolType.cs
namespace Atlasd.Battlenet
{
    public class ProtocolType
    {
        public enum Types : byte
        {
            // Descriptive Protocols:
            Game      = 0x01, // Diablo, StarCraft, WarCraft
            BNFTP     = 0x02, // FTP server for Game protocol
            Chat      = 0x03, // Human text-based Chat gateway
            Chat_Alt1 = 0x43,
            Chat_Alt2 = 0x63,
            IPC       = 0x80, // Inter-Process Communication

            // Letterized Protocols:
            Letter_C = 0x43,
            Letter_G = 0x47,
            Letter_H = 0x48,
            Letter_O = 0x4F,
            Letter_P = 0x50,
        };

        public Types Type { get; private set; }

        public ProtocolType(Types type)
        {
            Type = type;
        }

        public bool IsBNFTP()
        {
            return Type == Types.BNFTP;
        }

        public bool IsChat()
        {
            return
                Type == Types.Chat ||
                Type == Types.Chat_Alt1 ||
                Type == Types.Chat_Alt2
            ;
        }

        public bool IsGame()
        {
            return Type == Types.Game;
        }

        public bool IsInterProcessCommunication()
        {
            return Type == Types.IPC;
        }

        public bool IsIPC()
        {
            return IsInterProcessCommunication();
        }

        public static string ProtocolTypeName(Types type)
        {
            return type switch
            {
                Types.Game      => "Game",
                Types.BNFTP     => "BNFTP",
                Types.Chat      => "Chat",
                Types.Chat_Alt1 => "Chat_Alt1",
                Types.Chat_Alt2 => "Chat_Alt2",
                Types.IPC       => "IPC",
                _ => $"Unknown (0x{(byte)type:X2})",
            };
        }

        public override string ToString()
        {
            return ProtocolTypeName(Type);
        }
    }
}

File: src/AtlasV/Battlenet/BNET/ProtocolType.vb
Imports AtlasV.Daemon

Namespace AtlasV.Battlenet
    Public Class ProtocolType

        Public Enum Types As Byte
            Game = &H1
            BNFTP = &H2
            Chat = &H3
            Chat_Alt1 = &H43
            Chat_Alt2 = &H63
            IPC = &H80
            Letter_C = &H43
            Letter_G = &H47
            Letter_H = &H48
            Letter_O = &H4F
            Letter_P = &H50
        End Enum


        Public Property Type As Types

        Public Sub New(ByVal varType As Types)
            Type = varType
        End Sub

        Public Function IsBNFTP() As Boolean
            Return Type = Types.BNFTP
        End Function

        Public Function IsChat() As Boolean
            Return Type = Types.Chat OrElse Type = Types.Chat_Alt1 OrElse Type = Types.Chat_Alt2
        End Function

        Public Function IsGame() As Boolean
            Return Type = Types.Game
        End Function

        Public Function IsInterProcessCommunication() As Boolean
            Return Type = Types.IPC
        End Function

        Public Function IsIPC() As Boolean
            Return IsInterProcessCommunication()
        End Function

        Public Shared Function ProtocolTypeName(ByVal varType As Types) As String
            Select Case varType
                Case Types.Game : Return "Game"
                Case Types.BNFTP : Return "BNFTP"
                Case Types.Chat : Return "Chat"
                Case Types.Chat_Alt1 : Return "Chat_Alt1"
                Case Types.Chat_Alt2 : Return "Chat_Alt2"
                Case Types.IPC : Return "IPC"
                Case Else : Return $"Unknown (0x{CByte(varType)})"
            End Select
        End Function

        Public Shared Function ProtocolTypeToLogType(varType As Types) As UInt32
            Select Case varType
                Case Types.Game : Return Logging.LogType.Client_Game
                Case Types.BNFTP : Return Logging.LogType.Client_BNFTP
                Case Types.Chat : Return Logging.LogType.Client_Chat
                Case Types.Chat_Alt1 : Return Logging.LogType.Client_Chat
                Case Types.Chat_Alt2 : Return Logging.LogType.Client_Chat
                Case Types.IPC : Return Logging.LogType.Client_IPC
                Case Else : Return Logging.LogType.Client
            End Select
        End Function

        Public Overrides Function ToString() As String
            Return ProtocolTypeName(Type)
        End Function

    End Class
End Namespace
